### PR TITLE
Allow @API to be declared at the package level

### LINKED ISSUE #19: Allow @API to be declared at the package level
I am reviewing https://github.com/mockito/mockito/pull/2047 atm, which adds apiguardian statements to various Mockito internal classes that are used by the `mockito-junit-jupiter` extension. This approach is quite ad-hoc, as we would have to add the annotation when we start using it. An alternative approach would be to add the annotation to every single class we have in `org.mockito.internal`, but that seems to be quite invasive as well.

Would it be possible to allow the annotation to be added on package statements (e.g. in a `package-info.java`) for `org.mockito.internal`? That would then indicate that all classes in its (sub-)packages are marked internal.

I am not sure if OSGi tooling would need to be updated to support the new location, but at least it would allow us to more nicely integrate `apiguardian` into Mockito.

CC @rotty3000

### PR DESCRIPTION
**Pull request by TimvdLippe**
Fixes #19

### CODE REVIEW

File: src/main/java/org/apiguardian/api/API.java [hunk 1]
```diff
@@ -38,9 +39,13 @@
  * a class annotated with {@code @API(status = STABLE)} may declare a constructor
  * for internal usage that is annotated with {@code @API(status = INTERNAL)}.
  *
+ * <p>If {@code @API} is present on a package, it is considered to hold for all
+ * public types in its package and its subpackages. The same rules for lowered
```
Comment (marcphilipp) [inline src/main/java/org/apiguardian/api/API.java:43]: I'm not so sure about the subpackages part. In my experience, annotations have to declared on each package, e.g. for nullability.

@sbrannen WDYT?

Comment (sbrannen) [reply src/main/java/org/apiguardian/api/API.java:43]: Indeed. Java doesn't have support for "subpackages". In other words, there's no reflection-based support in the JDK for retrieving all subpackages of a given `java.lang.Package`. Similarly, there is no way to traverse up the package hierarchy using reflection like you can with type hierarchies (e.g., `java.lang.Class::getSuperclass()`). An annotation declared on a package is also not `@Inherited` by anything.

So, there is no built-in support for determining if a "super package" is annotated with a particular annotation.

One can of course handcraft an algorithm to navigate package hierarchies by parsing fully qualified package names.

In any case, I tend to agree with @marcphilipp in that we should not claim that an `@API` annotation declared on a package is valid for "subpackages".

Comment (TimvdLippe) [reply src/main/java/org/apiguardian/api/API.java]: Fixed.

**Review by marcphilipp — APPROVED** [review]

**Comment by TimvdLippe** [comment]
Friendly ping @sbrannen

**Comment by rotty3000** [comment]
Is there any chance this could get merged? It seems there is consensus? I've tested that this works as expected in bnd.

**Comment by sormuras** [comment]
Late, but not yet too late: let's drop the `Target` annotation.

Would be in-sync with https://github.com/junit-team/junit5/issues/2391

**Comment by marcphilipp** [comment]
> Late, but not yet too late: let's drop the `Target` annotation.

I think in this case we shouldn't do that since `@API` is a user-facing annotation (contrary to `@Testable`).

**Comment by marcphilipp** [comment]
> Is there any chance this could get merged? It seems there is consensus? I've tested that this works as expected in bnd.

@rotty3000 Merged and released in 1.1.1.

@TimvdLippe Thanks! 👍

**Pull request merged**